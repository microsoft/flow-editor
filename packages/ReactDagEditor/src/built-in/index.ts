export * from "./rect";
export * from "./line";
export * from "./defaultPort";
export * from "./InMemoryClipboard";
export * from "./InMemoryStorage";
export * from "./applyDefaultPortsPosition";
export * from "./applyHorizontalGraphPortsPosition";
