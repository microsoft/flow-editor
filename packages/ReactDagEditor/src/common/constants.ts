/**
 * https://developer.mozilla.org/en-US/docs/Web/API/MouseEvent/button
 */
export enum MouseEventButton {
  /**
   * usually the left button or the un-initialized state
   */
  Primary = 0,
  /**
   * usually the wheel button or the middle button (if present)
   */
  Auxiliary = 1,
  /**
   * usually the right button
   */
  Secondary = 2,
  /**
   * typically the Browser Back button
   */
  Fourth = 4,
  /**
   * typically the Browser Forward button
   */
  Fifth = 5,
}

export enum MouseEventButtons {
  None = 0,
  Left = 1,
  Right = 2,
  Middle = 4,
}

export const DEFAULT_AUTO_ALIGN_THRESHOLD = 50;
export const COPIED_NODE_SPACING = 50;
export const NODE_MIN_VISIBLE_LENGTH = 5;
export const NODE_MAX_VISIBLE_LENGTH = 500;
