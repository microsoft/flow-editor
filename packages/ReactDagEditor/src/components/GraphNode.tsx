/* eslint-disable react/jsx-no-bind */
import * as React from "react";
import { useTheme } from "../hooks";
import { GraphNodeEvent, INodeCommonEvent, INodeContextMenuEvent } from "../models/event";
import { IViewport } from "../models/geometry";
import { NodeModel } from "../models/NodeModel";
import { INodeConfig } from "../models/settings";
import { getNodeAutomationId, getNodeUid } from "../utils";
import { EventChannel } from "../utils/eventChannel";
import { checkIsMultiSelect } from "../utils/keyboard";
import classes from "./Graph.styles.m.scss";
import { IGraphProps } from "./Graph/IGraphProps";

export interface IGraphNodeCommonProps {
  isNodeEditDisabled?: boolean;
  eventChannel: EventChannel;
  viewport: Required<IViewport>;
  getNodeAriaLabel: Required<IGraphProps>["getNodeAriaLabel"];
  graphId: string;
}

export interface IGraphNodeProps extends IGraphNodeCommonProps {
  node: NodeModel;
  nodeConfig: INodeConfig;
}

const GraphNode: React.FunctionComponent<IGraphNodeProps> = props => {
  const { node, eventChannel, getNodeAriaLabel, viewport, graphId, nodeConfig } = props;
  const { theme } = useTheme();

  const nodeEvent = (type: (INodeCommonEvent | INodeContextMenuEvent)["type"]) => (
    evt: React.SyntheticEvent | React.MouseEvent
  ) => {
    evt.persist();
    const e = {
      type,
      node,
      rawEvent: evt
    } as INodeCommonEvent | INodeContextMenuEvent;
    eventChannel.trigger(e);
  };

  const clickEvent: React.MouseEventHandler = e => {
    e.persist();
    eventChannel.trigger({
      type: GraphNodeEvent.Click,
      rawEvent: e,
      isMultiSelect: checkIsMultiSelect(e),
      node
    });
  };

  const id = getNodeUid(graphId, node);

  const automationId = node.automationId ?? getNodeAutomationId(node);

  return (
    <g
      id={id}
      // for IE and Edge
      focusable="true"
      tabIndex={0} // why hard coded 0? - Dynamically calculating tabindex costs a lot in big graph. Then we always let it have tabIndex 0 to be focusable and override the "tab" key behavior.
      className={classes.node}
      onPointerDown={nodeEvent(GraphNodeEvent.PointerDown)}
      onPointerEnter={nodeEvent(GraphNodeEvent.PointerEnter)}
      onPointerMove={nodeEvent(GraphNodeEvent.PointerMove)}
      onPointerLeave={nodeEvent(GraphNodeEvent.PointerLeave)}
      onPointerUp={nodeEvent(GraphNodeEvent.PointerUp)}
      onDoubleClick={nodeEvent(GraphNodeEvent.DoubleClick)}
      onMouseDown={nodeEvent(GraphNodeEvent.MouseDown)}
      onMouseUp={nodeEvent(GraphNodeEvent.MouseUp)}
      onMouseEnter={nodeEvent(GraphNodeEvent.MouseEnter)}
      onMouseLeave={nodeEvent(GraphNodeEvent.MouseLeave)}
      onContextMenu={nodeEvent(GraphNodeEvent.ContextMenu)}
      onMouseMove={nodeEvent(GraphNodeEvent.MouseMove)}
      onMouseOver={nodeEvent(GraphNodeEvent.MouseOver)}
      onMouseOut={nodeEvent(GraphNodeEvent.MouseOut)}
      onKeyDown={nodeEvent(GraphNodeEvent.KeyDown)}
      onClick={clickEvent}
      aria-label={getNodeAriaLabel(node)}
      role="group"
      aria-roledescription="node"
      data-automation-id={automationId}
    >
      <g className="node-box-container">
        {nodeConfig.render({
          model: node,
          theme,
          viewport
        })}
      </g>
    </g>
  );
};

export { GraphNode };
