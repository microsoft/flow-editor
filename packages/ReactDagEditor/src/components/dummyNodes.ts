import { IDummyNodes } from "../models/dummy-node";

export const emptyDummyNodes = (): IDummyNodes => ({
  dx: 0,
  dy: 0,
  dWidth: 0,
  dHeight: 0,
  alignedDX: undefined,
  alignedDY: undefined,
  nodes: [],
  isVisible: false
});
