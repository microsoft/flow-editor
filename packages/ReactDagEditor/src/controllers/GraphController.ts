import { IDispatch } from "../contexts";
import { GraphFeatures } from "../Features";
import { IPoint } from "../models/geometry";
import { GraphModel } from "../models/GraphModel";
import { NodeModel } from "../models/NodeModel";
import { GraphBehavior, IGraphState } from "../models/state";
import { EventChannel } from "../utils/eventChannel";

/**
 * event handlers must get/set GraphBehavior immediately
 * to determine how to treat event objects (eg. preventDefault)
 * but dispatch/setState can be asynchronous
 * thus GraphBehavior is mirrored here
 */
export class GraphController {
  /**
   * since we don't have a detailed specification for touch handling
   * temporarily store the pointerId here for most single point events
   */
  public pointerId: number | null = null;
  /**
   * simulate canvas click event once
   */
  public canvasClickOnce = false;
  public nodeClickOnce: NodeModel | null = null;
  public readonly eventChannel = new EventChannel();
  public state: IGraphState;
  public UNSAFE_latestState: IGraphState;
  public dispatchDelegate: IDispatch;
  public getGlobalEventTargetImpl?: () => Window | Element | null | undefined;
  private mouseClientPoint?: IPoint;
  private behavior = GraphBehavior.default;

  public constructor(state: IGraphState, dispatch: IDispatch) {
    this.state = state;
    this.UNSAFE_latestState = state;
    this.dispatch = dispatch;
  }

  public readonly dispatch: IDispatch = (action, callback) => {
    this.dispatchDelegate(action, callback);
  };

  public setMouseClientPosition(pos: IPoint): void {
    this.mouseClientPoint = pos;
  }

  public unsetMouseClientPosition(): void {
    this.mouseClientPoint = undefined;
  }

  public getMouseClientPosition(): IPoint | undefined {
    return this.mouseClientPoint;
  }

  public getEnabledFeatures(): ReadonlySet<GraphFeatures> {
    return this.state.settings.features;
  }

  public getBehavior(): GraphBehavior {
    return this.behavior;
  }

  public setBehavior(value: GraphBehavior): void {
    this.behavior = value;
  }

  public getData(): GraphModel {
    return this.state.data.present;
  }

  public getGlobalEventTarget(): Element | Window {
    return this.getGlobalEventTargetImpl?.() ?? window;
  }
}
