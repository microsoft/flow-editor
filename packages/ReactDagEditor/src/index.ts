export * from "./components";
export * from "./contexts";
export * from "./built-in";
export * from "./utils";
export * from "./hooks";
export * from "./testAPI";
export * from "./Features";
export * from "./models/EdgeModel";
export * from "./models/GraphModel";
export * from "./models/NodeModel";
export * from "./reducers/composeReducers";
export * from "./models/element-state";
export * from "./models/port";
export * from "./models/node";
export * from "./models/edge";
export * from "./models/canvas";
export * from "./models/geometry";
export * from "./models/state";
export * from "./models/event";
