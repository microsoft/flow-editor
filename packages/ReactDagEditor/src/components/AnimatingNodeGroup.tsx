import * as React from "react";
import { defaultColors } from "../common/constants";
import { useGraphConfig } from "../hooks/context";
import type { IDummyNodes } from "../models/dummy-node";
import type { GraphModel } from "../models/GraphModel";
import { getNodeConfig } from "../utils";
import { Slots } from "./Slots/Slots";

interface IAnimatingNodeGroup {
  dummyNodes: IDummyNodes;
  graphData: GraphModel;
}

export const AnimatingNodeGroup: React.FunctionComponent<IAnimatingNodeGroup> = (props) => {
  const { dummyNodes, graphData } = props;
  const graphConfig = useGraphConfig();
  const { dWidth, dHeight } = dummyNodes;
  const dx = dummyNodes.alignedDX ?? dummyNodes.dx;
  const dy = dummyNodes.alignedDY ?? dummyNodes.dy;
  return (
    <g>
      {dummyNodes.nodes.map((dummyNode) => {
        const node = graphData.nodes.get(dummyNode.id);
        if (!node) {
          return null;
        }
        const x = dummyNode.x + dx;
        const y = dummyNode.y + dy;
        const width = dummyNode.width + dWidth;
        const height = dummyNode.height + dHeight;
        const nodeConfig = getNodeConfig(node, graphConfig);

        if (nodeConfig?.renderDummy) {
          return nodeConfig.renderDummy({
            ...node.inner,
            x,
            y,
            width,
            height,
          });
        }
        return (
          <Slots.NodeFrame height={height} width={width} key="node-frame">
            <rect
              key={node.id}
              transform={`translate(${x},${y})`}
              height={height}
              width={width}
              stroke={defaultColors.dummyNodeStroke}
              strokeDasharray="4"
              fill="none"
            />
          </Slots.NodeFrame>
        );
      })}
    </g>
  );
};
