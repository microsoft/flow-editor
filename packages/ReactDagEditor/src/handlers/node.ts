import * as React from "react";
import { MouseEventButton } from "../common/constants";
import { IGraphConfig } from "../contexts";
import { DragNodeController, TGetPositionFromEvent } from "../controllers";
import { PointerEventProvider } from "../event-provider/PointerEventProvider";
import { GraphNodeEvent, IEvent } from "../models/event";
import { IContainerRect } from "../models/geometry";
import { NodeModel } from "../models/NodeModel";
import { isWithinThreshold } from "../utils";
import { EventChannel } from "../utils/eventChannel";
import { graphController } from "../utils/graphController";
import { checkIsMultiSelect } from "../utils/keyboard";

export interface INodePointerDownParams {
  svgRef: React.RefObject<SVGSVGElement>;
  rectRef: React.RefObject<IContainerRect | undefined>;
  isNodesDraggable: boolean;
  isAutoAlignEnable: boolean;
  dragThreshold: number;
  getPositionFromEvent: TGetPositionFromEvent;
  isClickNodeToSelectDisabled: boolean;
  graphConfig: IGraphConfig;
  autoAlignThreshold: number;
  eventChannel: EventChannel;
}

export const onNodePointerDown = (evt: React.PointerEvent, target: NodeModel, params: INodePointerDownParams) => {
  evt.preventDefault();

  const {
    graphConfig,
    svgRef,
    isNodesDraggable,
    getPositionFromEvent,
    isClickNodeToSelectDisabled,
    eventChannel,
    dragThreshold,
    rectRef,
    isAutoAlignEnable,
    autoAlignThreshold
  } = params;

  if (isNodesDraggable) {
    evt.stopPropagation();
  }

  const isMouseRight = evt.pointerType === "mouse" && evt.button !== MouseEventButton.Primary;

  if (isClickNodeToSelectDisabled || isMouseRight) {
    return;
  }

  if (svgRef.current) {
    svgRef.current.focus({ preventScroll: true });
  }

  const isMultiSelect = checkIsMultiSelect(evt);

  const dragging = new DragNodeController(
    new PointerEventProvider(graphConfig.getGlobalEventTarget(), evt.pointerId),
    getPositionFromEvent,
    rectRef
  );

  dragging.onMove = ({ dx, dy, totalDX, totalDY, e }) => {
    if (isNodesDraggable) {
      eventChannel.trigger({
        type: GraphNodeEvent.Drag,
        node: target,
        dx,
        dy,
        rawEvent: e,
        isVisible: !isWithinThreshold(totalDX, totalDY, dragThreshold),
        isAutoAlignEnable,
        autoAlignThreshold
      });
    }
  };

  dragging.onEnd = ({ totalDX, totalDY, e }) => {
    graphController.pointerId = null;
    const events: IEvent[] = [];
    const isDragCanceled = isWithinThreshold(totalDX, totalDY, dragThreshold);
    events.push({
      type: GraphNodeEvent.DragEnd,
      node: target,
      rawEvent: e,
      isDragCanceled
    });
    if (isDragCanceled || !isNodesDraggable) {
      graphController.nodeClickOnce = target;
    }
    eventChannel.batch(events);
  };
  graphController.pointerId = evt.pointerId;
  if (evt.target instanceof Element && evt.pointerType !== "mouse") {
    evt.target.releasePointerCapture(evt.pointerId);
  }
  eventChannel.trigger({
    type: GraphNodeEvent.DragStart,
    node: target,
    rawEvent: evt,
    isMultiSelect
  });
  dragging.start(evt.nativeEvent);
};
