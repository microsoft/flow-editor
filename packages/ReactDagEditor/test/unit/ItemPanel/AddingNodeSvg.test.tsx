/* eslint-disable @typescript-eslint/no-explicit-any */
import { act, cleanup, fireEvent, render, RenderResult, screen } from "@testing-library/react";
import * as React from "react";
import * as ShallowRenderer from "react-test-renderer/shallow";
import { GraphCanvasEvent, GraphStateStore } from "../../../src";
import { MouseEventButton } from "../../../src/common/constants";
import { Item } from "../../../src/components/ItemPanel";
import { AddingNodeSvg } from "../../../src/components/ItemPanel/AddingNodeSvg";
import { GraphController } from "../../../src/controllers/GraphController";
import { ICanvasNode } from "../../../src/models/node";
import { GraphControllerRef } from "../../TestComponent";
import { mockClientRect, patchPointerEvent } from "../../utils";
import { withGraphConfigContext } from "../__mocks__/mockContext";
import { TestItemContent } from "./TestItemContent";

jest.mock("../../../src/components/ItemPanel/useSvgRect", () => ({
  useSvgRect: () => {
    return mockClientRect;
  }
}));

describe("ItemPanel - AddingNodeSvg", () => {
  let nodeWillAdd: () => ICanvasNode;
  let nodeDidAdd: () => void;
  let renderedWrapper: RenderResult;
  let graphController: GraphController;
  beforeAll(() => {
    patchPointerEvent();
    jest.useFakeTimers();
  });
  afterAll(() => {
    jest.clearAllMocks();
  });

  beforeEach(() => {
    nodeWillAdd = jest.fn();
    nodeDidAdd = jest.fn();
    const graphControllerRef = React.createRef<GraphController>();
    renderedWrapper = render(
      withGraphConfigContext(
        <GraphStateStore>
          <Item
            model={{ name: "node1", shape: "nodeShape" }}
            dragWillStart={jest.fn()}
            nodeWillAdd={nodeWillAdd}
            nodeDidAdd={nodeDidAdd}
          >
            <TestItemContent text="test item for addingNodeSVG" />
          </Item>
          <GraphControllerRef ref={graphControllerRef} />
        </GraphStateStore>
      )
    );
    graphController = graphControllerRef.current!;
    expect(graphController).toBeDefined();
    act(() => {
      graphController.dispatch({
        type: GraphCanvasEvent.ViewportResize,
        viewportRect: mockClientRect
      });
    });
  });

  afterEach(cleanup);

  /**
   * @param el
   * @param mouseEventButton
   * @param points
   */
  function simulateDragging(
    el: HTMLElement | Window,
    mouseEventButton: MouseEventButton,
    points: Array<[number, number]>
  ): void {
    if (points.length === 0) {
      return;
    }

    const first = points[0];
    act(() => {
      fireEvent.pointerDown(el, {
        clientX: first[0],
        clientY: first[1],
        button: mouseEventButton,
        pointerType: "mouse"
      });
      jest.runAllTimers();
    });

    points.forEach(([clientX, clientY]) => {
      act(() => {
        fireEvent.pointerMove(window, {
          clientX,
          clientY,
          button: mouseEventButton,
          pointerType: "mouse"
        });
      });
    });

    const last = points[points.length - 1];
    act(() => {
      jest.runAllTimers();
      fireEvent.pointerUp(window, {
        clientX: last[0],
        clientY: last[1],
        button: mouseEventButton,
        pointerType: "mouse"
      });
    });
  }

  it("Should match the snapshot", () => {
    const renderer = ShallowRenderer.createRenderer();
    renderer.render(
      <AddingNodeSvg
        nextNodeRef={{ current: null }}
        model={{
          name: "node1",
          shape: "nodeShape",
          x: 0,
          y: 0,
          id: "mock-id"
        }}
        svgRef={React.createRef<SVGSVGElement>()}
      />
    );
    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });

  it("adding node", () => {
    simulateDragging(screen.getByRole("button"), MouseEventButton.Primary, [
      [100, 100],
      [105, 105]
    ]);

    expect(nodeWillAdd).toBeCalledTimes(1);
    expect(nodeDidAdd).toBeCalledTimes(1);
    expect(renderedWrapper.container).toMatchSnapshot();
  });

  it("not adding node if out of graph", () => {
    jest.spyOn(Element.prototype, "getBoundingClientRect").mockReturnValue({
      top: 50,
      right: 200,
      bottom: 200,
      left: 50,
      width: 150,
      height: 150,
      x: 50,
      y: 50
    } as any);

    simulateDragging(screen.getByRole("button"), MouseEventButton.Primary, [
      [25, 25],
      [30, 30]
    ]);

    expect(nodeWillAdd).not.toBeCalled();
    expect(nodeDidAdd).not.toBeCalled();
    expect(renderedWrapper.container).toMatchSnapshot();
  });

  it("without left mouse down when dragging", () => {
    simulateDragging(screen.getByRole("button"), MouseEventButton.Secondary, [
      [100, 100],
      [105, 105]
    ]);

    expect(nodeWillAdd).not.toBeCalled();
    expect(nodeDidAdd).not.toBeCalled();
    expect(renderedWrapper.container).toMatchSnapshot();
  });
});
