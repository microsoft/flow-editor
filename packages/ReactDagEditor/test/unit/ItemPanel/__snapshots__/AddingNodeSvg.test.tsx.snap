// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemPanel - AddingNodeSvg Should match the snapshot 1`] = `
<svg
  className="adding-node-svg"
  id="mock-uuid"
  preserveAspectRatio="xMidYMid meet"
>
  <Transform
    matrix={
      Array [
        1,
        0,
        0,
        1,
        0,
        0,
      ]
    }
  >
    <GraphNode
      eventChannel={
        EventChannel {
          "externalHandlerRef": Object {
            "current": null,
          },
          "listenersRef": Object {
            "current": null,
          },
          "queue": Array [],
          "working": false,
        }
      }
      getNodeAriaLabel={[Function]}
      graphId="mock-uuid"
      node={
        Object {
          "id": "mock-id",
          "name": "node1",
          "shape": "nodeShape",
          "x": 0,
          "y": 0,
        }
      }
      viewport={
        Object {
          "rect": Object {
            "bottom": 700,
            "height": 600,
            "left": 100,
            "right": 900,
            "top": 100,
            "width": 800,
            "x": 100,
            "y": 100,
          },
          "transformMatrix": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
        }
      }
    />
  </Transform>
</svg>
`;

exports[`ItemPanel - AddingNodeSvg Should match the snapshot: events 1`] = `Array []`;

exports[`ItemPanel - AddingNodeSvg adding node 1`] = `
<div>
  <div
    class="module-item"
    role="button"
  >
    <p>
      test item for addingNodeSVG:default
    </p>
  </div>
</div>
`;

exports[`ItemPanel - AddingNodeSvg adding node: events 1`] = `Array []`;

exports[`ItemPanel - AddingNodeSvg not adding node if out of graph 1`] = `
<div>
  <div
    class="module-item"
    role="button"
  >
    <p>
      test item for addingNodeSVG:default
    </p>
  </div>
</div>
`;

exports[`ItemPanel - AddingNodeSvg not adding node if out of graph: events 1`] = `Array []`;

exports[`ItemPanel - AddingNodeSvg without left mouse down when dragging 1`] = `
<div>
  <div
    class="module-item"
    role="button"
  >
    <p>
      test item for addingNodeSVG:default
    </p>
  </div>
</div>
`;

exports[`ItemPanel - AddingNodeSvg without left mouse down when dragging: events 1`] = `Array []`;
